src/jit/gentree.cpp

```diff
@@ -14347,42 +14333,150 @@ bool GenTree::isCommutativeSIMDIntrinsic()
 }
 #endif //FEATURE_SIMD
 
-//-------------------------------------------------------------------------
-// Initialize: Return Type Descriptor given type handle.
+//---------------------------------------------------------------------------------------
+// InitializeStructReturnType:
+//    Initialize the Return Type Descriptor for a method that returns a struct type
 // 
 // Arguments
 //    comp        -  Compiler Instance
-//    retClsHnd   -  VM handle to the type returned
+//    retClsHnd   -  VM handle to the struct type returned by the method
 //
 // Return Value
 //    None
 //
-void ReturnTypeDesc::InitializeReturnType(Compiler* comp, CORINFO_CLASS_HANDLE retClsHnd)
+void ReturnTypeDesc::InitializeStructReturnType(Compiler* comp, CORINFO_CLASS_HANDLE retClsHnd)
 {
     assert(!m_inited);
 
-#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+#if FEATURE_MULTIREG_RET
+
     assert(retClsHnd != NO_CLASS_HANDLE);
+    unsigned  structSize = comp->info.compCompHnd->getClassSize(retClsHnd);
 
-    SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
-    comp->eeGetSystemVAmd64PassStructInRegisterDescriptor(retClsHnd, &structDesc);
+    Compiler::structPassingKind  howToReturnStruct;
+    var_types returnType = comp->getReturnTypeForStruct(retClsHnd, &howToReturnStruct, structSize);
 
-    if (structDesc.passedInRegisters)
+    switch (howToReturnStruct)
     {
-        for (int i=0; i<structDesc.eightByteCount; i++)
+        case Compiler::SPK_PrimitiveType:
         {
-            assert(i < MAX_RET_REG_COUNT);
-            m_regType[i] = comp->GetEightByteType(structDesc, i);
+            assert(returnType != TYP_UNKNOWN);
+            assert(returnType != TYP_STRUCT);
+            m_regType[0] = returnType;
+            break;
         }
-    }
 
-#elif defined(_TARGET_X86_)
-    // TODO-X86: Assumes we are only using ReturnTypeDesc for longs on x86.
-    // Will need to be updated in the future to handle other return types
-    assert(MAX_RET_REG_COUNT == 2);
+        case Compiler::SPK_ByValueAsHfa:
```

Comment:
Right, but that means that this will be dead code on x64/ux

---
In reply to: [72345619](https://github.com/dotnet/coreclr/pull/6467#discussion_r72345619) [](ancestors = 72345619)